nestedtensor/nested/masking.py

```diff
@@ -1,111 +1,184 @@
-The resulting NestedTensor will be of dimension n.
-Unless specific, it is attempted to yield a NestedTensor of minimal nested dimension.
-"""
-def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=None):
-    # TODO: Should be possible with views only.
-    if not mask.dim() > 0:
-        raise RuntimeError("Mask has to have dimention > 0")
-
-    # TODO: Need to define empty-list and 0 numel semantics
-    # TODO: Test edge-cases and kwargs
-    def _merge(tensors, nested_dim=None):
-        assert len(tensors) > 0
-
-        def _mergable_tensors(tensors, nested_dim):
-            if nested_dim is None:
-                size0 = tensors[0].size()
-                compliant = all(torch.is_tensor(t) for t in tensors)
-                compliant = all(size0 == t.size()
-                                for t in tensors) and compliant
-                return compliant
-            else:
-                return nested_dim < 1
 
-        if _mergable_tensors(tensors, nested_dim):
-            return torch.stack(tensors)
-        else:
-            return creation.nested_tensor(tensors)
+# Constructs nested tensor from passed tensor and mask.
+def nested_tensor_from_tensor_mask(tensor, mask, nested_dim=1):
+    if tensor is None:
+        raise RuntimeError("Tensor can't be undefined (None).")
+
+    if mask is None:
+        raise RuntimeError("Mask can't be undefined (None).")
+
+    # Scalar was passed
+    if tensor.dim() == 0:
+        raise RuntimeError("Can't construct nested tensor from a scalar.")
+
+    if nested_dim == 0:
+        raise RuntimeError("Nested dimension can't be 0.")
 
-    if nested_dim is not None and mask.dim() < nested_dim:
-        raise ValueError("Mask dimension ({0}) is too small to construct nested tensor with nested dimension of {1}".format(mask.dim(), nested_dim))
+    if nested_dim is not None and nested_dim > tensor.dim():
+        raise RuntimeError("Nested dimension ({0}) can't be bigger than data tensor dimension ({1}).".format(nested_dim, tensor.dim()))
 
-    if mask.dim() == 1:
-        tensors = [tensor[i] if mask[i]
-                   else None for i in range(len(mask))]
-        tensors = list(filter(lambda x: x is not None, tensors))
+    if tensor.numel() == 0 and mask.numel() != 0:
+        raise RuntimeError("Data tensor can't be emtpy if a mask has values.")
+
+    if tensor.numel() != 0 and mask.numel() == 0:
+        raise RuntimeError("Mask tensor can't be emtpy if a data tensor has values.")
+
+    return nt_from_tensor_mask(tensor, mask, nested_dim, tensor.dtype)
+
+
+def nt_from_tensor_mask(tensor, mask, nested_dim, dt):
+    def _merge(tensors, nested_dim):
         if len(tensors) == 0:
-            return tensor.narrow(0, 0, 0)
-        return _merge(tensors, nested_dim)
+            return torch.tensor([], dtype=dt)
+        return torch.stack(tensors)
+
+    if nested_dim == 0:
```

Comment:
I assume you want this for the recursive cases since you test against it in the caller. It might be more prudent to make this a private function, since without the error checking it supports nested_dim == 0, which we consider incorrect (i.e. doesn't return a NestedTensor).

Reply:
isn't `nt_from_tensor_mask` already ta private function?